### ISSUE TITLE
Question about the screenspace_points

### ISSUE THREAD

**Issue by Dingjifeng** [humansensinglab/taming-3dgs#27]
Thanks for your great work. But I have some issue with the code about the code.

in render.py
```
  screenspace_points = torch.zeros_like(pc.get_xyz, dtype=pc.get_xyz.dtype, requires_grad=True, device="cuda") + 0
    try:
        screenspace_points.retain_grad()
    except:
        pass
```

you create the  screenspace_points  from zeros.  and pass it to means2D:
```
 means3D = pc.get_xyz
 means2D = screenspace_points
 opacity = pc.get_opacity
```

But in the forward pass，the means2D is not used! so how do you obtain it's grads properly? 

![Image](https://github.com/user-attachments/assets/0924bc77-8fb6-4da7-8ca7-e17a16999e49)

![Image](https://github.com/user-attachments/assets/37019d2c-a5c7-4e7a-a32d-1888ff0ddd83)

Your help would be greateful!

**Comment by rahul-goel** [comment]
It is being passed in the forward function as a dummy tensor that is not used because PyTorch expects to have consistency between the input arguments in the forward function and the returned gradients in the backward function; otherwise it throws an error. Internally, it checks the shapes, datatypes, etc. of the gradient tensors and they should match the input tensors to the forward.

In the backward pass, the correct gradients are being returned as they are calculated and stored during gradient calculation, despite a dummy tensor in the forward.

**Comment by Dingjifeng** [comment]
> It is being passed in the forward function as a dummy tensor that is not used because PyTorch expects to have consistency between the input arguments in the forward function and the returned gradients in the backward function; otherwise it throws an error. Internally, it checks the shapes, datatypes, etc. of the gradient tensors and they should match the input tensors to the forward.
> 
> In the backward pass, the correct gradients are being returned as they are calculated and stored during gradient calculation, despite a dummy tensor in the forward.

Thanks for your reply。Sorry for have another Question which is about the max_contrib. 
```
CHECK_CUDA(FORWARD::render(
		tile_grid, block,
		imgState.ranges,
		binningState.point_list,
		imgState.bucket_offsets, sampleState.bucket_to_tile,
		sampleState.T, sampleState.ar,
		width, height,
		geomState.means2D,
		// feature_ptr,
		geomState.render_channels,
		geomState.conic_opacity,
		imgState.accum_alpha,
		imgState.n_contrib,
		imgState.max_contrib,
		background,
		out_color,
		contribCountBuffer,
		contribOffsetBuffer,
		imgState.contrib_scan,
		imgState.scan_size,
		listBuffer, listBufferRender, listBufferDistance,
		pixel_weights,
		accum_weights,
		reverse_count,
		blend_weights,
		dist_accum), debug)

	std::cout << "in forward:" << std::endl;
	uint32_t* max_contrib_cpu = (uint32_t*)malloc(sizeof(uint32_t) * 1000);
	cudaMemcpy(max_contrib_cpu,imgState.max_contrib,sizeof(uint32_t) * 1000,cudaMemcpyDeviceToHost);
	for(int i = 0; i < 1000; ++i){
		std::cout << max_contrib_cpu[i] << std::endl;
	}

	std::cout << "---" << std::endl;
```
I added some print function after forward. But the results are wired:

![Image](https://github.com/user-attachments/assets/327740d2-8809-4b87-8fce-a29bef31217b)

since i'am using truck dataset and the point cloud is 136029. So I'am confused if you encountered the same problem. Thank you!
